### PR TITLE
stops updatehealth() calls when there were no changes to a mob's health, cleans up a TON of redundant updatehealth() calls

### LINKED ISSUE #51299: Something is Erroneously Calling updatehealth every Few Seconds
Noticed then when glancing at cyborg code; how module unquipping is handled is in `updatehealth`: https://github.com/tgstation/tgstation/blob/master/code/modules/mob/living/silicon/robot/robot.dm#L706

This makes sense in the context of making something update the moment their health falls below a threshold.

That said, this made me raise an eyebrow; a long time ago @phil235 made `Life` conditionally update, so things like health overlays aren't reliant on a loop to function and only update when they need to.

Naturally this made me wonder "where's the code that handles when you re-equip an item to a slot on a borg? Does it just call `updatehealth` when it does so to make it unequip, or does it just not equip to begin with?

I couldn't find anything so I booted up to test; damaged myself as a borg until my modulde unequipped, then I re-equipped an item--after a bit of a delay it unequipped the item again.

After that I threw in some debugging text for `updatehealth` and rebooted---not shockingly, `updatehealth` was getting called roughly every 2 seconds, which is roughly how often `Life` takes place.

I have no idea if, for sure, it is `Life` or what is doing this, but the delay is roughly every 2 seconds.

This is an obvious performance problem, as this means that a lot of icons are being rendered (this is happening on every mob it seems) for no reason and is a violation of the system originally put into place.

### LINKED ISSUE #52468: AI Integrity Restorer functionality broken.
## Information

AI Integrity Restorer's functionality was broken in #52265 by accidentally removing updatehealth() calls.

image

Because AI Integrity Restorer functionality relies on health being updated, it will no longer revive a destroyed AI. It will heal a damaged AI, but the AI's health will only be updated when they took damage again after being placed back in their core.

#52391 fixes this unintentionally, by refactoring health update code which fixes this issue by changing code functionality.
#52466 fixes this directly.

Making the issue report to document the issue and the two PRs that resolve it.

### PR DESCRIPTION
**Pull request by ShizCalev**
Closes #52388
Closes #52392
Closes #51299
Fixes #52468
Closes #52466 
:cl: ShizCalev
fix: Removed some free lag caused by updatehealth() being called by damage procs even when a mob's health hasn't changed.
/:cl:

### CODE REVIEW

**Comment by MrMelbert** [comment]
image
minor concern- a few places in the code use ^ this method to get around calling updateHealth multiple times - would this pr fail to update if, say, brute was healed and oxygen was not?

File: code/modules/mob/living/damage_procs.dm [hunk 1]
```diff
@@ -169,15 +174,22 @@
...
 
 
 /mob/living/proc/setOxyLoss(amount, updating_health = TRUE, forced = FALSE)
 	if(!forced && status_flags & GODMODE)
 		return
+	if(oxyloss == amount)
+		return
 	. = oxyloss
 	oxyloss = amount
 	if(updating_health)
```
Comment (ShizCalev) [inline code/modules/mob/living/damage_procs.dm:195]: would be called every tick of Life() if the mob had TRAIT_NOBREATH

File: code/datums/status_effects/buffs.dm [hunk 2]
```diff
@@ -273,28 +267,15 @@
...
-		for(var/mob/living/L in view(7, owner))
+		for(var/mob/living/L in viewers(7, owner))
 			if(L.health < L.maxHealth)
 				new /obj/effect/temp_visual/heal(get_turf(L), "#375637")
-			if(iscarbon(L))
-				L.adjustBruteLoss(-3.5)
-				L.adjustFireLoss(-3.5)
-				L.adjustToxLoss(-3.5, forced = TRUE) //Because Slime People are people too
-				L.adjustOxyLoss(-3.5)
-				L.adjustStaminaLoss(-3.5)
```
Comment (ShizCalev) [inline code/datums/status_effects/buffs.dm:292]: >calling updatehealth() 5 times every tick for every mob on screen, reeeeeeee

**Comment by MrMelbert** [comment]
man I spent the last two days doing exactly this and you come in and do it in like 4 hours
this closes #52392 now
also be sure to get the fiery_rebirthspell in eldritch_magic.dm

**Comment by ShizCalev** [comment]
There's a lot more I have to hit, have to manually check everything that calls all the individual adjustdamage procs

File: code/modules/mob/living/carbon/damage_procs.dm [hunk 3]
```diff
@@ -74,22 +76,29 @@
...
 	if(HAS_TRAIT(src, TRAIT_TOXIMMUNE)) //Prevents toxin damage, but not healing
 		amount = min(amount, 0)
-	return ..()
+	. = ..()
+	if(isnull(.) || forced || !tox_immune)
+		return
+	if(amount > 0) //finish toxin lover damage if we actually changed health
+		blood_volume -= 5*amount
+	else
+		blood_volume -= amount
```
Comment (Dennok) [inline code/modules/mob/living/carbon/damage_procs.dm:101]: This make imposible to create blood for slimepeople by eating toxins if slimeperson at max health?

**Comment by Timberpoes** [comment]
Closes #52466 - However, I'm leaving 52466 open as AI restorer functionality is unique and right now the only method of restoring a dead AI is via admin intervention, so I'd consider it a high priority problem. I'm making an issue and linking both PRs to document it.

**Comment by AnturK** [comment]
Still WIP ?

**Comment by ShizCalev** [comment]
Will have this done by Saturday evening; been a busy week.

File: code/modules/mob/living/carbon/damage_procs.dm (hunk 3 above)
Comment (Rohesie) [reply code/modules/mob/living/carbon/damage_procs.dm:101]: Is this still an issue?

File: code/modules/reagents/chemistry/reagents/alcohol_reagents.dm [hunk 4]
```diff
@@ -517,12 +517,9 @@ All effects don't start immediately, but rather get worse over time; the rate is
...
 	if(M.mind && M.mind.has_antag_datum(/datum/antagonist/rev)) //Cuba Libre, the traditional drink of revolutions! Heals revolutionaries.
-		M.adjustBruteLoss(-1, 0)
-		M.adjustFireLoss(-1, 0)
-		M.adjustToxLoss(-1, 0)
-		M.adjustOxyLoss(-5, 0)
-		. = 1
-	return ..() || .
+		M.heal_overall_damage(brute = 1, burn = 1, toxin = 1, oxy = 5, updating_health = FALSE)
+		. = TRUE
+	..()
```
Comment (Rohesie) [inline code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:522]: Sure you want to ignore the parent return?

**Comment by Rohesie** [comment]
Wish `updating_health` was `FALSE` by default.
Let us know when this is ready.

**Comment by SpaceManiac** [comment]
PR has been marked WIP long enough that I'm closing as stale.

**Pull request closed without merging**